core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -290,7 +367,19 @@ class half_t {
 
   template <class T>
   KOKKOS_FUNCTION void operator=(T rhs) volatile {
-    val = cast_to_half(rhs).val;
+    uint16_t* val_ptr = (uint16_t*)const_cast<impl_type*>(&val);
+    impl_type new_val = cast_to_half(rhs).val;
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    alignas(4) float rhs_tmp = __half2float(new_val);
+    volatile float* new_lhs  = &rhs_tmp;
+    val                      = __float2half(new_lhs[0]);
+#else
+    Kokkos::atomic_store(val_ptr, *((uint16_t*)&new_val));
```

Comment:
call explicitly the CUDA atomic store/volatile assignment?

Reply:
cuda does not have atomic store. __half supports volatile assignment but not loads.

Reply:
Is this what you're referring to?

```c++
#include <cuda_fp16.h>

int main(int argc, char **argv) {
  __half h = 0.0;
  volatile __half vh = 1.0;
  h = *(&vh);
  *(&vh) = h;

  return 0;
}
```

```
$ nvcc -c nvcc_half_t_test.cc 
nvcc_half_t_test.cc: In function ‘int main(int, char**)’:
nvcc_half_t_test.cc:7:10: warning: implicit dereference will not access object of type ‘volatile __half’ in statement
   *(&vh) = h;
   ~~~~~~~^~~
```